src/ripple/app/hook/Macro.h

```diff
@@ -0,0 +1,241 @@
+
+
+
+
+#define COMPUTE_HOOK_DATA_OWNER_COUNT(state_count)\
+    (std::ceil( (double)state_count/(double)5.0 ))
+
+#define HOOK_SETUP()\
+    [[maybe_unused]] ApplyContext& applyCtx = hookCtx.applyCtx;\
+    [[maybe_unused]] auto& view = applyCtx.view();\
+    [[maybe_unused]] auto j = applyCtx.app.journal("View");\
+    [[maybe_unused]] unsigned char* memory = WasmEdge_MemoryInstanceGetPointer(&memoryCtx, 0, 0);\
+    [[maybe_unused]] const uint64_t memory_length = WasmEdge_MemoryInstanceGetPageSize(&memoryCtx) * \
+        WasmEdge_kPageSize;
+
+#define WRITE_WASM_MEMORY(bytes_written, guest_dst_ptr, guest_dst_len,\
+        host_src_ptr, host_src_len, host_memory_ptr, guest_memory_length)\
+{\
+    int64_t bytes_to_write = std::min(static_cast<int64_t>(host_src_len), static_cast<int64_t>(guest_dst_len));\
+    if (guest_dst_ptr + bytes_to_write > guest_memory_length)\
+    {\
+        JLOG(j.warn())\
+            << "HookError[" << HC_ACC() << "]: "\
+            << __func__ << " tried to retreive blob of " << host_src_len\
+            << " bytes past end of wasm memory";\
+        return OUT_OF_BOUNDS;\
+    }\
+    WasmEdge_MemoryInstanceSetData(&memoryCtx, \
+            reinterpret_cast<const uint8_t*>(host_src_ptr), guest_dst_ptr, bytes_to_write);\
+    bytes_written += bytes_to_write;\
+}
+
+#define WRITE_WASM_MEMORY_AND_RETURN(guest_dst_ptr, guest_dst_len,\
+        host_src_ptr, host_src_len, host_memory_ptr, guest_memory_length)\
+{\
+    int64_t bytes_written = 0;\
+    WRITE_WASM_MEMORY(bytes_written, guest_dst_ptr, guest_dst_len, host_src_ptr,\
+            host_src_len, host_memory_ptr, guest_memory_length);\
+    return bytes_written;\
+}
+
+#define RETURN_HOOK_TRACE(read_ptr, read_len, t)\
+{\
+    int rl = read_len;\
+    if (rl > 1024)\
+        rl = 1024;\
+    if (NOT_IN_BOUNDS(read_ptr, read_len, memory_length))\
+    {\
+        return OUT_OF_BOUNDS;\
+    }\
+    else if (read_ptr == 0 && read_len == 0)\
+    {\
+        JLOG(j.trace()) \
+            << "HookTrace[" << HC_ACC() << "]: " << t;\
+    }\
+    else if (is_UTF16LE(memory + read_ptr, rl))\
+    {\
+        uint8_t output[1024];\
+        int len = rl / 2;\
+        for (int i = 0; i < len && i < 512; ++i)\
+            output[i] = memory[read_ptr + i * 2];\
+        JLOG(j.trace()) \
+            << "HookTrace[" << HC_ACC() << "]: "\
+            << std::string_view((const char*)output, (size_t)(len)) << " "\
+            << t;\
+    }\
+    else\
+    {\
+        JLOG(j.trace()) \
+            << "HookTrace[" << HC_ACC() << "]: "\
+            << std::string_view((const char*)(memory + read_ptr), (size_t)rl) << " "\
+            << t;\
+    }\
+    return 0;\
+}
+// ptr = pointer inside the wasm memory space
+#define NOT_IN_BOUNDS(ptr, len, memory_length)\
+    ((static_cast<uint64_t>(ptr) > static_cast<uint64_t>(memory_length)) || \
+     ((static_cast<uint64_t>(ptr) + static_cast<uint64_t>(len)) > static_cast<uint64_t>(memory_length)))
```

Comment:
👍Good point

From memory the lower bound check was actually (although not obviously) to help check for an overflow, but the inputs are always going to be i32 from web assembly and it's an incomplete overflow check anyway. Since these are very inexpensive checks (fastest possible CPU operations) so I think we should go with a better defensive (not strictly necessary) overflow check instead of the first part of the || like so:

```c++
static_cast<uint64_t>(ptr) + static_cast<uint64_t>(len) < std::min(static_cast<uint64_t>(ptr), static_cast<uint64_t>(len))
```

Happy to hear your thoughts

Reply:
Maybe I am missing something, but I think this is equivalent to `a + b < min(a, b)` which, considering a and b are positive, will never be true. If the original i32 numbers can be negative maybe we can do a better check?

Reply:
If for whatever reason at some point this bounds check macro is used with uint64_t's (keeping in mind macro arguments are not typed checked) then it's best to check for overflow here as a defensive programming principle. a + b can be less than min(a,b) if a+b overflows the uint64_t.
Here's an illustration:
```c++
#include <stdint.h>
#include <iostream>
#include <algorithm>
int main() {

    auto overflowTest = [](uint64_t a, uint64_t b)
    {
        uint64_t c = a + b;
        std::cout <<
            "a=" << a << ", "
            "b=" << b << ". "
            "a+b=" << (a+b) << ". "
            "a + b < min(a,b) ? " << (a + b < std::min(a,b) ? "yes": "no") << "\n";
    };

    // without an overflow event
    overflowTest(0xFFFFFFFFFFFFFFFDULL, 2);
    // with an overflow event
    overflowTest(0xFFFFFFFFFFFFFFFEULL, 2);
    return 0;
}
```

Reply:
Well, if the inputs are always going to be i32 from webassembly, I don't think the check as described would serve any purpose.  I though  the `NOT_IN_BOUNDS` was intended to check whether we were indeed accessing memory withing the webassembly block.

Reply:
> Well, if the inputs are always going to be i32 from webassembly, I don't think the check as described would serve any purpose. I though the `NOT_IN_BOUNDS` was intended to check whether we were indeed accessing memory withing the webassembly block.

Agree. It's defensive: I.e. what happens if somehow for some reason uint64_ts are passed into the macro and they overflow the integer.